[TAN-5957] Fix bug where files couldn't be selected in file attachment widget

amanda-anderson opened the pull request:
# Changelog
## Fixed
- [TAN-5957] Fix bug where files couldn't be selected in file attachment widget.

amanda-anderson committed:
WIP create file attachment on layout save

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Merge branch 'TAN-5957-file-attachment-widget-bug-v3' of github.com:CitizenLabDotCo/citizenlab into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Code cleanup

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Merge branch 'TAN-5957-file-attachment-widget-bug-v3' of github.com:CitizenLabDotCo/citizenlab into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Fix unit test

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
WIP cleanup orphaned file attachments on save

amanda-anderson committed:
Simplify BE

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Code cleanup

amanda-anderson committed:
Code cleanup

amanda-anderson commented on back/engines/commercial/content_builder/app/services/content_builder/file_attachment_processor_service.rb:
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
```
@adessy Could I get your thoughts here? Is this approach ok with you?

I tried implementing a purely FE approach the last couple days, but this BE service feels way cleaner :thinking: All the different content builders hit this (project description builder/report builder/page builder), while in the FE the solution would need to be added to each one (and if we add more builders in the future, it will also need to be added).

I also ran into issues in the FE around comparing Files to File Attachments (since file attachments don't store the File ID by design, but then we would have to make separate requests to get the File info & File ID).

Since we already have some content builder layout processing, it feels like it fits well enough in our current code?

adessy replied:
Cool. I agree with the general approach. My main reservation is that we need to manually update the Craft.js state, which is not really a standard or public interface. So it needs to be well tested to catch any breaking changes. I'd suggest hooking the service directly into an AR callback so it's part of the same transaction.

amanda-anderson committed:
Small tweaks and add spec

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson commented on back/engines/commercial/content_builder/app/services/content_builder/file_attachment_processor_service.rb:
```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module ContentBuilder
```
Flagging the use of AI to assist in writing this service.

I've gone over it and made a few tweaks, and I think it looks good to me though.

amanda-anderson commented on back/engines/commercial/content_builder/spec/services/content_builder/file_attachment_processor_service_spec.rb:
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
```
Used AI to help write this spec as well. It originally over-tested though so I've scaled it back to these tests which I think make sense :)

amanda-anderson committed:
fix lint

adessy reviewed (commented):
Finally getting to this one! Sorry for the wait. I had a look before my holidays and had a few higher-level suggestions before doing a more in-depth review:
- I would suggest only using `fileId` and removing `fileAttachmentId` from the props of the `FileAttachment` widget. That way, we can skip the whole logic that updates the CraftJS state. The widget component could query all attachments for the given file ID and display the file if there is an attachment for the current layout.
- The layout ID can probably be found through the project (which is retrieved from the page params using `useParams` in some widgets), but I think using React context would be a better approach here.
- The current solution does not check whether the current user is authorized to create the file attachments. I think `LayoutPolicy#update?` (and `LayoutPolicy#create?`) should delegate to `FileAttachmentPolicy#create?` for each file being attached.

What do you think?

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson commented:
@adessy 

> 1) I would suggest only using fileId and removing fileAttachmentId from the props of the FileAttachment widget. That way, we can skip the whole logic that updates the CraftJS state. The widget component could query all attachments for the given file ID and display the file if there is an attachment for the current layout.

Do you mean to skip the logic that looks at the current CraftJS state to find what files are currently being used as attachments?

I don't think we can remove that and replace it with any sort of API call, since we still want to find any files that have been added in a file widget to the layout, but haven't been saved yet (I.e. these wont show up if we query all attachments for a given file ID)?

> 2) The layout ID can probably be found through the project

If we don't change anything from the #1 comment, we don't need to change anything here either right? Unless I've misunderstood something.

> 3) The current solution does not check whether the current user is authorized to create the file attachments.

Good point! I'll make that change :+1:

adessy commented:
> Do you mean to skip the logic that looks at the current CraftJS state to find what files are currently being used as attachments?
> 
> I don't think we can remove that and replace it with any sort of API call, since we still want to find any files that have been added in a file widget to the layout, but haven't been saved yet (I.e. these wont show up if we query all attachments for a given file ID)?

@amanda-anderson Could something along those lines work? It probably wouldn’t work as is, but it gives an idea of what I have in mind. The point is that if the widget props don’t include the attachment ID, we don’t need to update the CraftJS state manually anymore.

```javascript
const FileAttachment = ({ fileId, fileName }: FileAttachmentProps) => {
  // would be better to get the layout using a context provider
  const { projectId } = useParams();
  const { data: layout } = useContentBuilderLayout('project', projectId || '');

  const { data: attachments } = useFileAttachments({
    attachable_id: layout?.data.id,
    attachable_type: 'ContentBuilder::Layout',
  });

  const attachment = useMemo(() => {
    return attachments?.data.find(a => a.relationships.file.data.id === fileId);
  }, [attachments, fileId]);

  // Show placeholder with just the file name if we haven't saved yet
  if (!attachment) {
    if (fileName) {
      return (
        <Box
          id="e2e-file-attachment"
          maxWidth="1200px"
          margin="0 auto"
          style={{ pointerEvents: 'none' }}
        >
          <FileDisplay
            file={{
              id: 'temp-id',
              type: 'file',
              attributes: {
                name: fileName,
                file: { url: '' },
                ordering: null,
                size: 0,
                created_at: '',
                updated_at: '',
              },
            }}
          />
        </Box>
      );
    }
    return null;
  }

  // Saved state - show actual file
  return (
    <Box id="e2e-file-attachment" maxWidth="1200px" margin="0 auto">
      <FileDisplay
        file={{
          id: attachment.id,
          type: 'file',
          attributes: {
            ordering: attachment.attributes.position,
            name: attachment.attributes.file_name,
            size: attachment.attributes.file_size,
            created_at: attachment.attributes.created_at,
            updated_at: attachment.attributes.updated_at,
            file: {
              url: attachment.attributes.file_url,
            },
          },
        }}
      />
    </Box>
  );
};
```

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Code changes from code review

amanda-anderson committed:
Improve performance

amanda-anderson committed:
Remove deadcode

amanda-anderson committed:
Fix type issue

amanda-anderson commented on back/engines/commercial/content_builder/app/policies/content_builder/layout_policy.rb:
```diff
@@ -2,6 +2,12 @@
 
 module ContentBuilder
```
Copilot made this change for me - it seems it was required to properly return the attachments in the useFileAttachments call for the layouts.

adessy committed:
[TAN-5957] Add `extract_file_ids` class method to `FileAttachmentProcessorService`

adessy committed:
[TAN-5957] Add file authorization check to `LayoutPolicy#update?`

When updating a layout with file attachments in craftjs_json, verify
the user is authorized to use those files via FilePolicy#update?.

adessy committed:
[TAN-5957] Remove `fileAttachmentId` from the `FileAttachment` widget

Previously, `fileAttachmentId` was used to fetch attachments directly via `useFileAttachmentById`. This has been replaced with a fileId-based lookup. Rationale:
- `fileAttachmentId` was redundant since `fileId` already uniquely identifies the file, and attachments can be found by matching against it.
- This avoids patching craftjs_json on the backend after creating attachments, as we should avoid manually modifying this internal representation when possible.

Older layouts may still have `fileAttachmentId` stored in their craftjs_json. This is harmless since the prop is now ignored. A migration could be added later to clean up these stale values if needed.

adessy committed:
[TAN-5957] Move file attachment sync logic from service to Layout model

adessy committed:
[TAN-5957] Filter out missing files before creating attachments

adessy committed:
[TAN-5957] Improve FileAttachment widget UX for missing files

adessy committed:
[TAN-5957] Extract `FilePlaceholder` to separate file

adessy committed:
[TAN-5957] Rename `FileAttachmentEditor` to `FilePreview`

adessy committed:
[TAN-5957] Use distinct icon for empty FileAttachment placeholder

Use file-add icon for empty state to improve visual distinction between unconfigured and configured widgets.

adessy committed:
[TAN-5957] Improve FileAttachment placeholder copy

adessy committed:
[TAN-5957] Remove unnecessary code

adessy committed:
[TAN-5957] Add missing file authorization check to `LayoutPolicy#upsert?`

amanda-anderson committed:
Small tweaks FE

Commit:
Translations updated by CI (extract-intl)

amanda-anderson committed:
BE tweak to layout policy

amanda-anderson committed:
Merge branch 'TAN-5957-file-attachment-widget-bug-v3--adrien2' of github.com:CitizenLabDotCo/citizenlab into TAN-5957-file-attachment-widget-bug-v3--adrien2

amanda-anderson committed:
Merge pull request #12876 from CitizenLabDotCo/TAN-5957-file-attachment-widget-bug-v3--adrien2

Suggestions for "[TAN-5957] Fix bug where files couldn't be selected in file attachment widget"

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson committed:
Merge branch 'master' into TAN-5957-file-attachment-widget-bug-v3

amanda-anderson commented:
Issue causing E2E test failures already on master, so this doesn't block my release.

amanda-anderson commented:
(PR was approved by Adrien in DM).
